ConsistencyManager/DataStructures/WeakModelUpdatesListenerArray.swift

```diff
@@ -0,0 +1,170 @@
+     */
+    public init(arrayLiteral elements: ConsistencyManagerUpdatesListener?...) {
+        data = []
+        for element in elements {
+            data.append(weakClosureWithValue(element))
+        }
+    }
+
+    /**
+     Private initializer which allows init with data.
+     */
+    private init(data: [() -> ConsistencyManagerUpdatesListener?]) {
+        self.data = data
+    }
+
+    // MARK: Public Properties
+
+    /// How many elements the Array stores
+    public var count: Int {
+        return data.count
+    }
+
+    // MARK: Public Methods
+
+    /**
+     Append an element to the array.
+     */
+    public mutating func append(_ element: ConsistencyManagerUpdatesListener?) {
+        data.append(weakClosureWithValue(element))
+    }
+
+    /**
+     This method iterates through the array and removes any element which is nil.
+     It also returns an array of nonoptional values for convenience.
+
+     This method runs in O(n), so you should only call this method every time you need it. You should only call it once.
+     i.e. Don't do this:
+     for _ in array.prune()
+     */
+    public mutating func prune() -> [ConsistencyManagerUpdatesListener] {
+        var nonOptionalElements = [ConsistencyManagerUpdatesListener]()
+        data = data.filter { closure in
+            let value = closure()
+            if let value = value {
+                nonOptionalElements.append(value)
+                return true
+            } else {
+                return false
+            }
+        }
+        return nonOptionalElements
+    }
+
+    public func map(_ function: (ConsistencyManagerUpdatesListener?) -> ConsistencyManagerUpdatesListener?) -> WeakModelUpdatesListenerArray {
+        var newArray = WeakModelUpdatesListenerArray()
+        // TODO: Fix this once apple fixes their bug
+        // This currently crashes with EXC_BAD_ACCESS
+        //    for value in self {
+        //      let newValue = function(value)
+        //      newArray.append(newValue)
+        //    }
+        for closure in data {
+            let newValue = function(closure())
+            newArray.append(newValue)
+        }
+        return newArray
+    }
+
+    /**
+     Iterates over the collection with an include element closure.
+     It will remove any elements that return false.
+     */
+    func filter(_ includeElement: (ConsistencyManagerUpdatesListener?) -> Bool) -> WeakModelUpdatesListenerArray {
+        let newData = data.filter { closure in
+            return includeElement(closure())
+        }
+        return WeakModelUpdatesListenerArray(data: newData)
+    }
+
+    // MARK: Private Methods
+
+    fileprivate func weakClosureWithValue(_ object: ConsistencyManagerUpdatesListener?) -> () -> ConsistencyManagerUpdatesListener? {
+        return { [weak object] in
+            return object
+        }
+    }
+}
+
+// MARK: MutableCollectionType Implementation
+
+extension WeakModelUpdatesListenerArray: MutableCollection {
```

Comment:
I recommend taking a look at [this article](https://swiftunboxed.com/open-source/CollectionOfOne/?utm_campaign=This%2BWeek%2Bin%2BSwift&utm_medium=email&utm_source=This_Week_in_Swift_113) for info on implementing the protocol. I think you could do it in less code that simply calls down into the underlying array for all implementation.